4.x: Add optional fallback for `ControlConnection#reconnect() `

Bouncheck opened the pull request:
Adds additional config option that changes behavior of ControlConnection's reconnection. In case LBP returns an empty query plan we will attempt to connect to the original endpoints regardless of their last observed status.

The option is by default disabled and driver without enabling it should keep behaving the same way as before. 
If enabled it should be possible to reconnect to the cluster if all endpoints containing hostnames were lost during normal operation and then cluster was replaced.

Fix: https://github.com/scylladb/java-driver/issues/224

Bouncheck committed:
Add `MockResolverIT#cannot_reconnect_with_resolved_socket()`

Adds a method for testing the issues that surface after cluster
replacements. Due to the variable, sometimes long runtime it is not added
to any of the test groups.

dkropachev commented on integration-tests/src/test/java/com/datastax/oss/driver/core/resolver/MockResolverIT.java:
```diff
@@ -251,6 +253,161 @@ public void run_replace_test_20_times() {
     }
   }
 
+  @Test
```
Can you please mark this test as reproduction to make sure we don't run it

Bouncheck committed:
Add optional fallback for `ControlConnection#reconnect()`

Adds an experimental option to allow `ControlConnection` to try
reconnecting to the original contact points held by `MetadataManager`,
in case of getting empty query plan from the load balancing policy.

In order to separate this logic from query plans of other queries
`LoadBalancingPolicyWrapper#newControlReconnectionQueryPlan()` was introduced
and is called during reconnection in place of `newQueryPlan()`.

dkropachev commented:
@Bouncheck , I have been thinking about this solution, more I think less I like it.
This solution to feed hosts to controll connection when it reconnects looks hacky to me. 
I think driver should sync initial contact points to the list of nodes on node events (UP/DOWN/ADD/DEL).
Let's discuss it.

dkropachev commented on core/src/main/java/com/datastax/oss/driver/api/core/config/DefaultDriverOption.java:
```diff
@@ -668,6 +668,15 @@ public enum DefaultDriverOption implements DriverOption {
+   * <p>Value-type: boolean
+   */
+  CONTROL_CONNECTION_RECONNECT_CONTACT_POINTS(
+      "advanced.control-connection.reconnection.fallback-to-original-contacts"),
```
```suggestion
      "advanced.control-connection.reconnection.fallback-to-original-contact-points"),
```

Bouncheck replied:
Adjusted as suggested

dkropachev commented on integration-tests/src/test/java/com/datastax/oss/driver/core/resolver/MockResolverIT.java:
```diff
@@ -251,6 +255,201 @@ public void run_replace_test_20_times() {
+          break;
+        }
+      }
+      /*
+      for (int i = 0; i < 15; i++) {
+        try {
+          nodes = session.getMetadata().getNodes().values();
+          if (nodes.size() == 3) {
+            break;
+          }
+          Thread.sleep(1000);
+        } catch (InterruptedException e) {
+          break;
+        }
+      }
+       */
```
Do you need this code ?

Bouncheck replied:
Removed it

dkropachev replied:
Still see it.

Bouncheck replied:
Reappeared during rebase, removed it again

dkropachev commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/LoadBalancingPolicyWrapper.java:
```diff
@@ -165,6 +166,28 @@ public Queue<Node> newQueryPlan() {
+  @NonNull
+  public Queue<Node> newControlReconnectionQueryPlan() {
+    // First try the original way
+    Queue<Node> regularQueryPlan = newQueryPlan();
```
I believe you can drop `newQueryPlan();` at all

Bouncheck replied:
Why? Don't we want the driver to choose the hosts as policy dictates as long as the cluster is healthy?

dkropachev replied:
I just mean that instead of calling `newQueryPlan()` to call `newQueryPlan(null, DriverExecutionProfile.DEFAULT_NAME, null)` here

Bouncheck replied:
Replaced that call with the explicit one

dkropachev commented:
@Bouncheck , I see that you still use `newQueryPlan` at one particular place, can you please switch to `newControlReconnectionQueryPlan`
```
        Queue<Node> nodes = context.getLoadBalancingPolicyWrapper().newQueryPlan();
```

dkropachev commented:
> @Bouncheck , I have been thinking about this solution, more I think less I like it. This solution to feed hosts to controll connection when it reconnects looks hacky to me. I think driver should sync initial contact points to the list of nodes on node events (UP/DOWN/ADD/DEL). Let's discuss it.

Regarding this, I have tried to do it, unfortunately due to the API this fix will be extremely intrusive, let's go ahead with your solution.

Bouncheck commented:
> @Bouncheck , I see that you still use `newQueryPlan` at one particular place, can you please switch to `newControlReconnectionQueryPlan`
> 
> ```
>         Queue<Node> nodes = context.getLoadBalancingPolicyWrapper().newQueryPlan();
> ```

Yes, but it's during the initialization. I believe that part works fine.

dkropachev commented:
> > @Bouncheck , I see that you still use `newQueryPlan` at one particular place, can you please switch to `newControlReconnectionQueryPlan`
> > ```
> >         Queue<Node> nodes = context.getLoadBalancingPolicyWrapper().newQueryPlan();
> > ```
> 
> Yes, but it's during the initialization. I believe that part works fine.

It works fine, but creates unnecessary confusion.

Bouncheck commented:
I reverted that 1 call in ControlConnection init back to `newQueryPlan()`. I forgot that removing it would require changing around 20 test methods which expect that exact api call

Bouncheck commented:
I can adjust them, but I don't know if it's worth it

dkropachev commented:
> I can adjust them, but I don't know if it's worth it

Please check last commit and tell if you are ok with it.
